proposals/2023/04-21_build-init-log.md

```diff
@@ -0,0 +1,292 @@
+
+## Design
+
+<!--
+This section is intended to explain the solution design for the proposal.
+
+NOTE: If there are no current plans for a solution, please leave this section blank.
+-->
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+<!-- Answer here -->
+
+### Types
+
+No changes required for `database.Log` or `library.Log` structs since they are already associated with a `BuildID`.
+Validation, however, will need to allow for both `StepID` and `ServiceID` to be `NULL`.
+
+Drop this check in `Validate()` in `database/log.go` since no `StepID` and no `ServiceID` means we have a `Build` `Log`:
+<!-- https://github.com/go-vela/types/blob/main/database/log.go#L129-L132 -->
+```
+	// verify the has StepID or ServiceID field populated
+	if l.StepID.Int64 <= 0 && l.ServiceID.Int64 <= 0 {
+		return ErrEmptyLogStepOrServiceID
+	}
+```
+<!-- https://github.com/go-vela/types/blob/main/database/log_test.go#L298-L299
+And this needs to be failure: false
+```
+		{ // no service_id or step_id set for log
+			failure: true,
+```
+-->
+
+No changes are needed in the `pipeline` or the `yaml` layers because there is not a `pipeline.Log` nor a `yaml.Log`, so the relationship between `Build` and `Log` does not need to be represented.
+
+### Server
+
+#### Server - Database
+
+Adjust `CreateLog`, `UpdateLog`, and `DeleteLog` to handle the build `Log` (adding a `logger.Tracef` entry and an `Error` specific to the Build Log).
+
+Add `database/log/get_build.go` with:
+```
+// GetLogForBuild gets a log by build ID from the database.
+func (e *engine) GetLogForBuild(b *library.Build) (*library.Log, error) {
+```
+
+See if we can make the build log sort before the step logs in `ListLogsForBuild`:
+<!-- https://github.com/go-vela/server/blob/main/database/log/list_build.go#L37-L40 -->
+```
+	err = e.client.
+		Table(constants.TableLog).
+		Where("build_id = ?", b.GetID()).
+		Order("step_id ASC").
+```
+
+Sqlite sorts with `NULLS FIRST` by default, and Postgres sorts with `NULLS LAST` (see: [How Are NULLs Sorted by Default?](https://learnsql.com/blog/how-to-order-rows-with-nulls/)).
+So this query is inconsistent between databases. And, the service logs returned by this query are not sorted. It would be nice to have this method return the logs sorted in a consistent manner.
+I would like to see the Build Log, then Step Logs sorted by step_id, then Service Logs sorted by service_id. Any suggestions on how to do that with gorm?
+
+We might need some kind of index or constraint on the `Log` table so that rows with NULL step_id and NULL service_id must have a unique build_id. But, I'm not sure how to create such an index/constraint.
```

Comment:
Maybe we try to cover this in the Go code with [`log.Validate()`](https://github.com/go-vela/types/blob/12031186b20d769fd105ade236e8cc41f7e706df/database/log.go#L126-L145) rather than through the DB?

https://pkg.go.dev/github.com/go-vela/types/database#Log.Validate

Reply:
Are you suggesting that we query the database in `log.Validate()` to see if there are any other instances with NULL step_id and NULL service_id?